packages/clay-shared/src/useFocusManagement.ts

```diff
@@ -0,0 +1,224 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {useEffect, useRef} from 'react';
+
+type FocusManager = {
+	key: string;
+	portal?: boolean;
+	value: HTMLElement;
+};
+
+const TAB_KEY_CODE = 9;
+const FOCUSABLE_ELEMENTS = [
+	'a[href]',
+	'[contenteditable]',
+	'[tabindex]:not([tabindex^="-"])',
+	'area[href]',
+	'button:not([disabled]):not([aria-hidden])',
+	'embed',
+	'iframe',
+	'input:not([disabled]):not([type="hidden"]):not([aria-hidden])',
+	'object',
+	'select:not([disabled]):not([aria-hidden])',
+	'textarea:not([disabled]):not([aria-hidden])',
+];
+
+export function useFocusManagement(loop: boolean = false) {
+	const focusManagerRef = useRef<Array<FocusManager>>([]);
+	const lastPositionRef = useRef<number | null>(null);
+	const nextElementOutsideScopeRef = useRef<HTMLElement | null>(null);
+
+	const createScope = (ref: any, key: string, portal: boolean = false) => {
```

Comment:
One nitpick I might add is changing the name away from `key` since that is typically used when talking about the key prop in react. Maybe we could go with `identifier` or `name`?

Reply:
Hm yeah, I think we can switch to `identifier` I like that.